### ISSUE TITLE
Questions of several implementation details

### ISSUE THREAD

**Issue by xlnwel** [ku2482/fqf-iqn-qrdqn.pytorch#9]
Hello @ku2482

May I ask you several implementation details and why you made these decisions?

1. In [this line](https://github.com/ku2482/fqf-iqn-qrdqn.pytorch/blob/72e48a085dcb213e58960ce00fc7347d54aaba5a/fqf_iqn_qrdqn/agent/fqf_agent.py#L169), you compute signs by comparing `sa_quantiles[I]` with `sa_quantiles[i-1]`(except the first one). Why don't you use `values_1>0` as the signs?
2. In [this line](https://github.com/ku2482/fqf-iqn-qrdqn.pytorch/blob/72e48a085dcb213e58960ce00fc7347d54aaba5a/fqf_iqn_qrdqn/network.py#L61), you initialize the weights of the FractionProposalNetwork using Xavier initialize with `gain=0.01`. What makes you choose this initialization?

**Comment by ku2482** [comment]
Hi @xlnwel 

1
When F_Z^{-1}(w) is non-decreasing, the sign is +1 and gradients are the same as Proposition 1.

However, when F_Z^{-1}(w) is decreasing, the assumption of Proposition 1 doesn't hold anymore.
In this case, I correct the gradients setting the sign to -1.

2
> The weights of the fraction proposal network are initialized so that initial probabilities are uniform as
in QR-DQN

It's hand-designed by me. The author mentioned above in the paper, so I made the weight smaller so that initial probabilities are closer to the uniform distribution.

Thanks :)

**Comment by xlnwel** [comment]
Hi @ku2482 

Thank you for reply :-)

1. I get why you use sign, but I'm wondering why not use `values_1>0` as the sign as it directly compare `sa_quantiles` against `sa_quantile_hats[:, :-1]` to ensure non-decreasing.
2. That was a smart choice as I found normal `xavier_uniform` can easily result in distribution collapse even if the distribution looks uniform at the beginning of the training.

**Comment by ku2482** [comment]
Hi @xlnwel 

> I get why you use sign, but I'm wondering why not use values_1>0 as the sign as it directly compare sa_quantiles against sa_quantile_hats[:, :-1]

I want to evaluate if F_Z^{-1}(w) is non-decreasing (or not) when tau_i < w < tau_{i+1}.
So I defined the sign as "F_Z^{-1}(tau_{i+1}) - F_Z^{-1}(tau_i)". (See the proof for Proposition 1.)

However, I can't calculate F_Z^{-1}(0.0) and F_Z^{-1}(1.0), so I instead use F_Z^{-1}(tau_hats[0]) and F_Z^{-1}(tau_hats[-1]) respectively.

Does it make sense?

**Comment by xlnwel** [comment]
Hi @ku2482,

Thank you for your explanation and patience :-) 

Perhaps because of my poor expression, it seems that you still have not got my idea- -. Yeah, your way definitely makes sense to say that if F_Z is non-decreasing(i.e., `F_Z^{-1}(tau_{i+1}) - F_Z^{-1}(tau_i)`), the sign is True. What I tried to express was that now that we computed `value` as `sa_quantiles - sa_quantile_hats[:, :-1]`(i.e., `F_Z^{-1}(tau_{i}) - F_Z^{-1}(\hat tau_{i-1})`), why not directly use `sa_quantiles > sa_quantile_hats[:, :-1]`(i.e., `F_Z^{-1}(tau_{i}) > F_Z^{-1}(\hat tau_{i-1})`) as the sign which seem more directly related to `value`?

**Comment by xlnwel** [comment]
Hi @ku2482,

Thank you for your explanation and patience :-) 

Perhaps because of my poor expression, it seems that you still have not got my idea- -. Yeah, your way definitely makes sense to say that if F_Z is non-decreasing(i.e., `F_Z^{-1}(tau_{i+1}) - F_Z^{-1}(tau_i)`), the sign is True. What I tried to express was that now that we computed `value` as `sa_quantiles - sa_quantile_hats[:, :-1]`(i.e., `F_Z^{-1}(tau_{i}) - F_Z^{-1}(\hat tau_{i-1})`), why not directly use `sa_quantiles > sa_quantile_hats[:, :-1]`(i.e., `F_Z^{-1}(tau_{i}) > F_Z^{-1}(\hat tau_{i-1})`) as the sign which seem more directly related to `value`?

**Comment by xlnwel** [comment]
Hi @ku2482,

Thank you for your explanation and patience :-) 

Perhaps because of my poor expression, it seems that you still have not got my idea- -. Yeah, your way definitely makes sense to say that if F_Z is non-decreasing(i.e., `F_Z^{-1}(tau_{i+1}) - F_Z^{-1}(tau_i)`), the sign is True. What I tried to express was that now that we computed `value` as `sa_quantiles - sa_quantile_hats[:, :-1]`(i.e., `F_Z^{-1}(tau_{i}) - F_Z^{-1}(\hat tau_{i-1})`), why not directly use `sa_quantiles > sa_quantile_hats[:, :-1]`(i.e., `F_Z^{-1}(tau_{i}) > F_Z^{-1}(\hat tau_{i-1})`) as the sign which seem more directly related to `value`?

**Comment by ku2482** [comment]
Hi @xlnwel 

> why not directly use sa_quantiles > sa_quantile_hats[:, :-1](i.e., F_Z^{-1}(tau_{i}) > F_Z^{-1}(\hat tau_{i-1})) as the sign

That's not what I want to calculate. It's not the same amount as what I've used.

**Comment by xlnwel** [comment]
Hi @ku2482 

But it also attempts to make sure the non-decreasing condition on F_Z -- `tau_{i} > \hat tau_{i-1}` -> `F_Z^{-1}() > F_Z^{-1}(\hat tau_{i-1})` -- am I right? Why do you think this is infeasible ?

**Comment by ku2482** [comment]
Hi, please see the proof for Proposition 1 carefully.

**Comment by xlnwel** [comment]
Hi @ku2482. 

Thank you. Are you interested in participating in the [Procgen challenge](https://www.aicrowd.com/challenges/neurips-2020-procgen-competition/leaderboards) organized by OpenAI? I'm currently in the 5th position of round 1. If you're interested and have time, maybe we can do this together.

**Comment by ku2482** [comment]
Hi @xlnwel 

Thank you for inviting :)
It would be great to work together, however, I've already joined the team to help my friend learn RL.

I hope we'll be able to work together next time.
Please feel free to contact me through [LinkedIn](https://www.linkedin.com/in/ku2482/).

Thanks.

**Comment by xlnwel** [comment]
Hi @ku2482 

No problem. What's your position right now? Maybe we can share some thoughts after the challenge is completed.
